### PR TITLE
CodeGen: Remove UsesMSVCFloatingPoint from MachineModuleInfo

### PR DESCRIPTION
**Pull request by arsenm**
This is only used by x86 and only used in the AsmPrinter module pass. I
think implementing this by looking at the underlying IR types instead
of the selected instructions is a pretty horrifying implementation,
but it's still available in the AsmPrinter.

This is https://reviews.llvm.org/D123933 resurrected.

I still don't know what the point of emitting _fltused is, but this
approach of looking at the IR types probably isn't the right way to
do this in the first place. If the intent is report any FP instructions,
this will miss any implicitly introduced ones during codegen. Also don't
know why just unconditionally emitting it isn't an option.

The last review mentioned the ARMs might want to emit this, but I'm
not going to go fix that. If someone wants to emit this on ARM, they
can move this to a common helper or analysis somewhere.

### CODE REVIEW

**Comment by arsenm** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/100368?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#100369** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/100369?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#100368** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/100368?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#100357** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/100357?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#100353** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/100353?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#100333** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/100333?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#100319** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/100319?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @arsenm and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by RKSimon — APPROVED** [review]
LGTM - cheers

**Comment by compnerd** [comment]
`_fltused` is used to identify the need for supporting functions and is used during the linking process. Always emitting results in code size increases and overlinking, but catching the generated ones is definitely something that we should be doing.

**Comment by arsenm** [comment]
> `_fltused` is used to identify the need for supporting functions and is used during the linking process. Always emitting results in code size increases and overlinking, but catching the generated ones is definitely something that we should be doing.

So really it should be looking for post-legalize calls to specific runtime libcalls

**Comment by compnerd** [comment]
Not entirely - `_fltused` indicates that any floating point operations are being used. The problem is that the kernel mode code does not permit floating point operations (though you can explicitly use `KeSaveFloatingPointState` and `KeRestoreFloatingPointState` to explicitly use floating point code in a region). The `_fltused` is a protection mechanism for that, with the user space C++ runtime ("ABI library" - a la libc++abi) support library (vcruntime) providing a default definition of the symbol. That way, if the user accidentally uses floating point code in the kernel, it is identified at link time and prevents the errant use. Doing this post legalization makes sense though.

**Review by rnk — APPROVED** [review]
I think @jyknight touched this last, and I may have advised him to simply use MMI to store this, since historically it has been used as a side channel for passing information along the code generation pipeline. It's good to avoid that, but it was what we were doing at the time. :)

I think the reason the existing scan was done as part of SDISel is that it is a scan over the IR, and ISel is the point at which we walk the IR to build the mostly-standalone MIR representation. Moving this scan to AsmPrinting is a bit distasteful, but it works and is less code. You could also make a cache-locality argument that walking the IR again during AsmPrinting is slow, but that's pretty weak.

Scanning the IR for FP types is obviously a gross approximation of `_fltused`. If we want to try to implement this properly, maybe we could scan all MachineInstrs during AsmPrinting for usage of vector or x87 FP operands. We can leave that out of scope, though.

In conclusion, we should land this as is, but I wanted to share some context.

File: llvm/lib/Target/X86/X86AsmPrinter.cpp [hunk 1]
```diff
@@ -975,6 +976,28 @@ static void emitNonLazyStubs(MachineModuleInfo *MMI, MCStreamer &OutStreamer) {
   }
 }
 
+/// True if this module is being built for windows/msvc, and uses floating
+/// point.  This is used to emit an undefined reference to _fltused.
```
Comment (rnk) [inline llvm/lib/Target/X86/X86AsmPrinter.cpp:980]: Please add a FIXME or TODO about computing this from MIR instead of the IR, and include some context about how this is used in Windows kernel or driver contexts to find and prevent code from modifying non-GPR registers.

**Comment by jyknight** [comment]
In particular, I think it would be incorrect to just look at floating-point operations at the MIR level, because calling printf with a constant floating point argument doesn't necessarily use any floating point instructions.

**Comment by arsenm** [comment]
### Merge activity

* **Jul 26, 3:21 AM EDT**: @arsenm started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/100368).
